src/main/java/org/dbos/apiary/mysql/MysqlConnection.java

```diff
@@ -0,0 +1,186 @@
+        validationLock.lock();
+        boolean valid = true;
+        for (String table: writtenKeys.keySet()) {
+            for (String key : writtenKeys.get(table)) {
+                // Has the key been modified by a transaction not in the snapshot?
+                Set<Long> writes = committedWrites.getOrDefault(table, Collections.emptyMap()).getOrDefault(key, Collections.emptySet());
+                for (Long write : writes) {
+                    if (write >= txc.xmax || activeTransactions.contains(write)) {
+                        valid = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if (valid) {
+            for (String collection: writtenKeys.keySet()) {
+                for (String key : writtenKeys.get(collection)) {
+                    committedWrites.putIfAbsent(collection, new ConcurrentHashMap<>());
+                    committedWrites.get(collection).putIfAbsent(key, ConcurrentHashMap.newKeySet());
+                    committedWrites.get(collection).get(key).add(txc.txID);
+                }
+            }
+        }
+        validationLock.unlock();
+        if (valid) {
+            String query = "";
+            try {
+                Connection c = ds.getConnection();
+                Statement s = c.createStatement();
+                for (String table : writtenKeys.keySet()) {
+                    for (String key : writtenKeys.get(table)) {
+                        query = String.format("UPDATE %s SET %s = %d WHERE %s = '%s' AND %s < %d AND %s = %d", table, MysqlContext.endVersion, txc.txID, MysqlContext.apiaryID, key, MysqlContext.beginVersion, txc.txID, MysqlContext.endVersion, Long.MAX_VALUE);
+                        s.execute(query);
+                    }
+                }
+                s.close();
+                c.close();
+            } catch (Exception e) {
+                e.printStackTrace();
+                logger.error("Failed to update valid txn {}", txc.txID);
+                logger.info("Validate update query: {}", query);
+            }
+        }
+        return valid;
+    }
+
+    @Override
+    public void garbageCollect(Set<TransactionContext> activeTransactions) {
+        long globalxmin = activeTransactions.stream().mapToLong(i -> i.xmin).min().getAsLong();
+        // No need to keep track of writes that are visible to all active or future transactions.
+        committedWrites.values().forEach(i -> i.values().forEach(w -> w.removeIf(txID -> txID < globalxmin)));
+        // Delete old versions that are no longer visible to any active or future transaction.
+        String query = "";
+        try {
+            Connection c = ds.getConnection();
+            Statement s = c.createStatement();
+            for (String tableName : committedWrites.keySet()) {
+                query = String.format("DELETE FROM %s WHERE %s < %d", tableName, MysqlContext.endVersion, globalxmin);
```

Comment:
Same as above.

Reply:
I do not think we can use a single SQL to delete from multiple tables at the same time. Note that this loop is over tableNames.

Reply:
Makes sense.  And I don't think it matters as well since this is not on the commit critical path.